Godeps/_workspace/src/github.com/docker/distribution/registry/storage/layerwriter.go

```diff
@@ -79,8 +79,26 @@ func (lw *layerWriter) Finish(dgst digest.Digest) (distribution.Layer, error) {
 
 	}
 
-	if err := lw.moveLayer(canonical); err != nil {
-		// TODO(stevvooe): Cleanup?
+	// On NFS filesystem file attributes are cached on client. File may
+	// not exist (os.Stat failes with NotExist) few seconds after the write
+	// has been completed.
+	for retries := 0; ; retries++ {
+		err := lw.moveLayer(canonical)
+		if err == nil {
+			if retries > 0 {
+				ctxu.GetLogger(lw.layerStore.repository.ctx).Debugf("(*layerWriter).Finish: layer moved after %d failed attempts", retries)
+			}
+			break
+		}
+		switch err.(type) {
+		case storagedriver.PathNotFoundError:
```

Comment:
this is concerning... I assume we move the file successfully, then get a PathNotFoundError attempting to verify the destination file? If we get that error, it seems unsafe to call moveLayer again. Wouldn't that be likely to fail with a different error (or worse, a *different* PathNotFoundError since it can't find the source file, if we already moved it)?